### PR TITLE
QGIS Project temporal settings

### PR DESCRIPTION
**Pull request by Samweli**
This PR add project time settings functionality, it includes work from this [PR](https://github.com/qgis/QGIS/pull/34497) which has added support for constrains of temporal layers from WMS-T provider.

User will be able to set the project temporal range by either using manual input or calculate it from the current projects temporal layers by clicking the provided "Calculate from Layers" button.

This implementation set the stage for adding animation feature for temporal layers in QGIS.

Example of how setting project temporal range will be.
set project time settings

Prior discussions about QGIS temporal support can be found [here](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/161) and [here ](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/128)

### CODE REVIEW

**Comment by nyalldawson** [comment]
Actually can you rebase off https://github.com/qgis/QGIS/pull/34619 ? It includes some minor fixes and documentation improvements

**Comment by Samweli** [comment]
> Actually can you rebase off #34619 ? It includes some minor fixes and documentation improvements

Ok

File: src/core/raster/qgsrasterlayer.cpp [hunk 1]
```diff
@@ -577,7 +579,9 @@ void QgsRasterLayer::init()
   mLastViewPort.mWidth = 0;
   mLastViewPort.mHeight = 0;
 
-  mTemporalProperties = qgis::make_unique<QgsRasterLayerTemporalProperties>();
+  // Initialize temporal properties
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayer.cpp]: Revert this change

File: src/core/raster/qgsrasterlayer.cpp [hunk 2]
```diff
@@ -945,11 +949,6 @@ bool QgsRasterLayer::ignoreExtents() const
   return mDataProvider ? mDataProvider->ignoreExtents() : false;
 }
 
-QgsRasterLayerTemporalProperties *QgsRasterLayer::temporalProperties()
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayer.cpp]: Revert this change

File: src/core/raster/qgsrasterlayer.cpp [hunk 3]
```diff
@@ -1998,9 +1997,6 @@ bool QgsRasterLayer::writeXml( QDomNode &layer_node,
     layer_node.appendChild( noData );
   }
 
-  // write temporal properties
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayer.cpp]: Was dropping this deliberate?

File: src/core/raster/qgsrasterlayer.h [hunk 4]
```diff
@@ -52,14 +51,43 @@ typedef QList < QPair< QString, QColor > > QgsLegendColorList;
 
 /**
  * \ingroup core
+ *  This class provides qgis with the ability to render raster datasets
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayer.h]: Revert the changes to this file

File: src/core/raster/qgsrasterlayertemporalproperties.h [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #include "qgis_sip.h"
 #include "qgsrange.h"
 #include "qgsmaplayertemporalproperties.h"
+#include "qgsrasterdataprovider.h"
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayertemporalproperties.h]: Some from https://github.com/qgis/QGIS/commit/313b000147d28bb7eea788774890af62c1c60b9f were reverted here, can you please fix?

File: src/gui/raster/qgsrasterlayertemporalpropertieswidget.h [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    /**
+     * Constructor for QgsRasterLayerTemporalPropertiesWidget.
+     */
+    QgsRasterLayerTemporalPropertiesWidget( QWidget *parent = nullptr, QgsMapLayer *layer = nullptr );
+
+    /**
+     * Sets the map canvas associtated with this temporal widget.
+     *
+     */
+    void setMapCanvas( QgsMapCanvas *canvas );
```
Comment (nyalldawson) [inline src/gui/raster/qgsrasterlayertemporalpropertieswidget.h]: I don't think the canvas is used here anymore, can you remove it from the API?

File: src/ui/CMakeLists.txt [hunk 7]
```diff
@@ -31,6 +31,7 @@ QT5_WRAP_UI(QGIS_UIS_H
   ${LABELING_UIS}
   ${NUMERICFORMAT_UIS}
   ${ATTRIBUTEFORMCONFIG_UIS}
+  ${TEMPORALWIDGET_UIS}
```
Comment (nyalldawson) [inline src/ui/CMakeLists.txt]: Shouldn't be required

File: src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+ ***************************************************************************/
+
+#include "qgsrasterdataprovidertemporalcapabilities.h"
+
+QgsRasterDataProviderTemporalCapabilities::QgsRasterDataProviderTemporalCapabilities( bool enabled )
+  :  QgsDataProviderTemporalCapabilities( enabled )
+{
+}
+
+//QgsRasterDataProviderTemporalProperties &QgsRasterDataProviderTemporalProperties::operator=( const QgsRasterDataProviderTemporalProperties &other )
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp]: remove

File: src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+{
+  if ( !isActive() )
+    setIsActive( true );
+
+  // Don't set temporal range outside fixed temporal range limits,
+  // instead set equal to the fixed temporal range
+  QDateTime begin;
+  QDateTime end;
+
+  if ( dateTimeRange.begin() < mFixedRange.begin() )
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp]: Should you check if mFixedRange.isValid() here ? What if a provider doesn't have a fixed range?

**Comment by nyalldawson** [comment]
Thanks for the updates @Samweli !

**Comment by nirvn** [comment]
@Samweli , great work, super exciting to see QGIS gaining temporal awareness in its core.

On the UI front, I'd ask you to move the new temporal configuration widgets away from the project properties' general tab into its own new temporal tab. We don't want the general tab to become the dumping ground for all new feature :wink: It'd also be more future-proof as I suspect we'll add more temporal-related settings in there.

File: src/ui/qgsprojectpropertiesbase.ui [hunk 10]
```diff
@@ -819,6 +813,139 @@
...
+                  <item>
+                   <widget class="QGroupBox" name="startDateTimeGroupBox">
+                    <property name="title">
+                     <string>Start date</string>
+                    </property>
+                    <layout class="QVBoxLayout" name="verticalLayout_10">
+                     <item>
+                      <layout class="QHBoxLayout" name="horizontalLayout_21">
+                       <item>
+                        <widget class="QDateTimeEdit" name="mStartDateTimeEdit">
```
[the remark this answers is not in the document]
Comment (Samweli) [reply src/ui/qgsprojectpropertiesbase.ui]: Ok

File: src/app/qgsprojectproperties.cpp [hunk 11]
```diff
@@ -2475,6 +2497,74 @@ void QgsProjectProperties::mButtonAddColor_clicked()
   mTreeProjectColors->addColor( newColor, QgsSymbolLayerUtils::colorToName( newColor ) );
 }
 
+void QgsProjectProperties::setEndAsStartButton_clicked()
```
[the remark this answers is not in the document]
Comment (Samweli) [reply src/app/qgsprojectproperties.cpp]: It may be useful when we want to show temporal status of different single date time instant layers using the instant set from the project time settings.

File: src/app/qgsprojectproperties.cpp [hunk 12]
```diff
@@ -2475,6 +2497,74 @@ void QgsProjectProperties::mButtonAddColor_clicked()
...
+  QgsRasterLayer *rasterLayer = nullptr;
+  QList< QDateTime > dates;
+
+  for ( QMap<QString, QgsMapLayer *>::const_iterator it = mapLayers.constBegin(); it != mapLayers.constEnd(); ++it )
+  {
+    currentLayer = it.value();
+
+    // Checking for WMS-T layers only, as of time writing this, they have full temporal support.
+    if ( currentLayer->type() == QgsMapLayerType::RasterLayer &&
+         currentLayer->providerType() ==    "wms" )
```
[the remark this answers is not in the document]
Comment (Samweli) [reply src/app/qgsprojectproperties.cpp]: If we remove this, that means even other raster layers ( non WMS-T) will all also be getting their temporal capabilities/properties from their data providers, but the current temporal API allows for the raster layers to have temporal properties and right now there is no implementation supporting other non WMS-T layers.

File: src/core/qgsmaplayer.h [hunk 13]
```diff
@@ -39,6 +39,7 @@
 #include "qgsmaplayerstyle.h"
 #include "qgsreadwritecontext.h"
 #include "qgsdataprovider.h"
+#include "qgsmaplayertemporalproperties.h"
```
[the remark this answers is not in the document]
Comment (Samweli) [reply src/core/qgsmaplayer.h]: Sure

File: src/core/raster/qgsrasterdataprovidertemporalcapabilities.h [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+     * \see setFixedReferenceTemporalRange()
+    */
+    const QgsDateTimeRange &fixedReferenceTemporalRange() const;
+
+    /**
+     * Sets the time enabled status.
+     *
+     * \see isTimeEnabled()
+     */
+    void setEnableTime( bool enabled );
```
[the remark this answers is not in the document]
Comment (Samweli) [reply src/core/raster/qgsrasterdataprovidertemporalcapabilities.h]: This is used to set the status of whether the time part of the temporal range should be used when querying layer data corresponding the temporal range.

File: src/core/raster/qgsrasterlayer.cpp (hunk 3 above)
Comment (Samweli) [reply src/core/raster/qgsrasterlayer.cpp]: Not it wasn't

File: src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp (hunk 9 above)
Comment (Samweli) [reply src/core/raster/qgsrasterdataprovidertemporalcapabilities.cpp]: Yap

**Comment by Samweli** [comment]
> Thanks for the updates @Samweli !

Thanks too for the review @nyalldawson .

**Comment by Samweli** [comment]
> @Samweli , great work, super exciting to see QGIS gaining temporal awareness in its core.
> 
> On the UI front, I'd ask you to move the new temporal configuration widgets away from the project properties' general tab into its own new temporal tab. We don't want the general tab to become the dumping ground for all new feature  It'd also be more future-proof as I suspect we'll add more temporal-related settings in there.

Thanks @nirvn!, sounds great, I ll move it to the new tab.

**Comment by Samweli** [comment]
@nirvn @nyalldawson , a new temporal tab after working on your reviews.
new_project_temporal_settings_look

**Comment by nirvn** [comment]
@Samweli , fantastic, much cleaner.

One more suggestion: I'd use a non-collapsible groupbox here now that you've moved settings on its own tab.

**Comment by Samweli** [comment]
> @Samweli , fantastic, much cleaner.
> 
> One more suggestion: I'd use a non-collapsible groupbox here now that you've moved settings on its own tab.

Sure @nirvn, done!

**Review by nyalldawson — COMMENTED** [review]
Can you also add some specific tests for the project time settings class? I'd like to
- test that the signal is emitted correctly when the time range is changed (and not emitted if the time range is set to the existing value)
- test the saving/loading time settings occurs correctly when a project is saved and loaded

You could use test_qgsprojectviewsettings.py as a good starting point for these.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Looks good, just need the CI fixed

**Comment by timlinux** [comment]
Thanks @Samweli and @nyalldawson

**Comment by pigreco** [comment]
Please add `Temporal` label
